# **Lab 02: Khai thác và trực quan hóa dữ liệu**

### **Danh sách các thành viên**

<center>

|         Họ tên         |   MSSV   |
| :--------------------- | :------: |
| Trần Nguyễn Nhật Cường | 22127048 |
|    Nguyễn Công Tuấn    | 22127436 |
|     Trần Đăng Tuấn     | 22127438 |

</center>

## **Cài đặt và sử dụng các thư viện cần thiết**

In [5]:
%pip install matplotlib pandas seaborn kaggle scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [6]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import zipfile
import os
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import StandardScaler

pd.options.display.float_format = "{:.2f}".format

**`matplotlib`** được dùng để vẽ biểu đồ trong Python. Có thể dùng để vẽ các loại biểu đồ như
histogram, biểu đồ tán xạ (scatter plot), biểu đồ đường (line chart) và biểu đồ cột (bar chart),...

**`pandas`** được dùng để xử lý và phân tích dữ liệu dạng bảng và cung cấp cấu trúc dữ liệu như
DataFrame và Series để dễ thao tác

**`seaborn`** dùng để vẽ biểu đồ dựa trên thư viện matplotlib nhưng biểu đồ sẽ trông đẹp hơn và cũng hỗ trợ các loại biểu đồ thống kê như violin plot, box plot, heatmap

**`sklearn (scikit-learn)`** là thư viện học máy phổ biến trong Python, được sử dụng để thực hiện các tác vụ học máy và xử lý dữ liệu. Trong đồ án này, các thành phần sau của sklearn được sử dụng:
- **`sklearn.cluster.KMeans`**: Cung cấp thuật toán KMeans để gom nhóm dữ liệu thành các cụm dựa trên sự tương đồng, được sử dụng để phân loại các nền kinh tế thành các nhóm.
- **`sklearn.metrics.silhouette_score`**: Dùng để tính điểm Silhouette, một chỉ số đánh giá chất lượng của việc gom nhóm, giúp xác định số lượng cụm tối ưu bằng cách đo lường mức độ tách biệt giữa các cụm.
- **`sklearn.preprocessing.StandardScaler`**: Dùng để chuẩn hóa dữ liệu, đưa các giá trị về cùng thang đo (trung bình bằng 0 và độ lệch chuẩn bằng 1), đảm bảo các chỉ số kinh tế có trọng số công bằng khi áp dụng thuật toán gom nhóm.

## **Thu thập dữ liệu**

*Nền tảng thu thập* - *Kaggle*

*Tên bộ dữ liệu thu thập* - Student Depression Dataset: Analyzing Mental Health Trends and Predictors Among Students

*Thông tin về bộ dữ liệu* - Bộ dữ liệu này được tổng hợp từ nhiều nguồn khác nhau nhằm mục đích phân tích và dự đoán mức độ trầm cảm ở sinh viên. Nó được thiết kế phục vụ nghiên cứu trong các lĩnh vực tâm lý học, khoa học dữ liệu và giáo dục. Bộ dữ liệu cung cấp những hiểu biết quan trọng về các yếu tố góp phần vào các vấn đề sức khỏe tâm thần của sinh viên, từ đó hỗ trợ xây dựng các chiến lược can thiệp sớm và cải thiện sức khỏe tinh thần cho đối tượng này.

In [7]:
!kaggle datasets download -d adilshamim8/student-depression-dataset -p ./

Dataset URL: https://www.kaggle.com/datasets/adilshamim8/student-depression-dataset
License(s): apache-2.0




  0%|          | 0.00/456k [00:00<?, ?B/s]
100%|██████████| 456k/456k [00:01<00:00, 442kB/s]
100%|██████████| 456k/456k [00:01<00:00, 442kB/s]


In [8]:
zip_file_path = './student-depression-dataset.zip'

with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall('./')

os.remove(zip_file_path) 
csv_file_path = os.path.join('./', 'student_depression_dataset.csv') 

## **Đọc bộ dữ liệu**

In [64]:
X = pd.read_csv(csv_file_path)

In [65]:
X.head()

id  Gender   Age           City Profession  Academic Pressure  \
0   2    Male 33.00  Visakhapatnam    Student               5.00   
1   8  Female 24.00      Bangalore    Student               2.00   
2  26    Male 31.00       Srinagar    Student               3.00   
3  30  Female 28.00       Varanasi    Student               3.00   
4  32  Female 25.00         Jaipur    Student               4.00   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0           0.00  8.97                2.00              0.00   
1           0.00  5.90                5.00              0.00   
2           0.00  7.03                5.00              0.00   
3           0.00  5.59                2.00              0.00   
4           0.00  8.13                3.00              0.00   

        Sleep Duration Dietary Habits   Degree  \
0          '5-6 hours'        Healthy  B.Pharm   
1          '5-6 hours'       Moderate      BSc   
2  'Less than 5 hours'        Healthy       BA   
3          '7-8 hours'       Moderate      BCA   
4          '5-6 hours'       Moderate   M.Tech   

  Have you ever had suicidal thoughts ?  Work/Study Hours Financial Stress  \
0                                   Yes              3.00              1.0   
1                                    No              3.00              2.0   
2                                    No              9.00              1.0   
3                                   Yes              4.00              5.0   
4                                   Yes              1.00              1.0   

  Family History of Mental Illness  Depression  
0                               No           1  
1                              Yes           0  
2                              Yes           0  
3                              Yes           1  
4                               No           0

## **Tiền xử lý dữ liệu**

In [66]:
for col in X.columns:
    print(f"Unique values in {col}:")
    print(X[col].unique())
    print("-" * 50)

Unique values in id:
[     2      8     26 ... 140689 140690 140699]
--------------------------------------------------
Unique values in Gender:
['Male' 'Female']
--------------------------------------------------
Unique values in Age:
[33. 24. 31. 28. 25. 29. 30. 27. 19. 20. 23. 18. 21. 22. 34. 32. 26. 39.
 35. 42. 36. 58. 49. 38. 51. 44. 43. 46. 59. 54. 48. 56. 37. 41.]
--------------------------------------------------
Unique values in City:
['Visakhapatnam' 'Bangalore' 'Srinagar' 'Varanasi' 'Jaipur' 'Pune' 'Thane'
 'Chennai' 'Nagpur' 'Nashik' 'Vadodara' 'Kalyan' 'Rajkot' 'Ahmedabad'
 'Kolkata' 'Mumbai' 'Lucknow' 'Indore' 'Surat' 'Ludhiana' 'Bhopal'
 'Meerut' 'Agra' 'Ghaziabad' 'Hyderabad' 'Vasai-Virar' 'Kanpur' 'Patna'
 'Faridabad' 'Delhi' 'Saanvi' 'M.Tech' 'Bhavna' "'Less Delhi'" 'City'
 '3.0' "'Less than 5 Kalyan'" 'Mira' 'Harsha' 'Vaanya' 'Gaurav' 'Harsh'
 'Reyansh' 'Kibara' 'Rashi' 'ME' 'M.Com' 'Nalyan' 'Mihir' 'Nalini'
 'Nandini' 'Khaziabad']
----------------------------------

### **Xử lý dữ liệu bất hợp lý**

In [67]:
valid_cities = {
    'visakhapatnam', 'bangalore', 'srinagar', 'varanasi', 'jaipur', 'pune', 'thane',
    'chennai', 'nagpur', 'nashik', 'vadodara', 'kalyan', 'rajkot', 'ahmedabad',
    'kolkata', 'mumbai', 'lucknow', 'indore', 'surat', 'ludhiana', 'bhopal',
    'meerut', 'agra', 'ghaziabad', 'hyderabad', 'vasai-virar', 'kanpur', 'patna',
    'faridabad', 'delhi'
}

def clean_city(city):
    city = str(city).strip().lower().replace("'", "").split()[0]  

    if city in valid_cities:
        return city.title() 
    else:
        return "Unknown" 

X["City"] = X["City"].apply(clean_city)

### **Định dạng các giá trị của các thuộc tính**

In [68]:
X['Sleep Duration'] = X['Sleep Duration'].apply(lambda x: str(x).replace("'", ""))
X['Profession'] = X['Profession'].apply(lambda x: str(x).replace("'", ""))
X['Degree'] = X['Degree'].apply(lambda x: str(x).replace("'", ""))

### **Xử lý các giá trị bị thiếu**

In [69]:
print(f"Unique values in Financial Stress:")
print(X['Financial Stress'].unique())

Unique values in Financial Stress:
['1.0' '2.0' '5.0' '3.0' '4.0' '?']


In [70]:
X['Financial Stress'] = X['Financial Stress'].replace('?', np.nan).astype(float)
median_value = X['Financial Stress'].median()
X['Financial Stress'] = X['Financial Stress'].fillna(median_value)

### **Tổng quan về dữ liệu sau khi tiền xử lý**

In [71]:
for col in X.columns:
    print(f"Unique values in {col}:")
    print(X[col].unique())
    print("-" * 50)

Unique values in id:
[     2      8     26 ... 140689 140690 140699]
--------------------------------------------------
Unique values in Gender:
['Male' 'Female']
--------------------------------------------------
Unique values in Age:
[33. 24. 31. 28. 25. 29. 30. 27. 19. 20. 23. 18. 21. 22. 34. 32. 26. 39.
 35. 42. 36. 58. 49. 38. 51. 44. 43. 46. 59. 54. 48. 56. 37. 41.]
--------------------------------------------------
Unique values in City:
['Visakhapatnam' 'Bangalore' 'Srinagar' 'Varanasi' 'Jaipur' 'Pune' 'Thane'
 'Chennai' 'Nagpur' 'Nashik' 'Vadodara' 'Kalyan' 'Rajkot' 'Ahmedabad'
 'Kolkata' 'Mumbai' 'Lucknow' 'Indore' 'Surat' 'Ludhiana' 'Bhopal'
 'Meerut' 'Agra' 'Ghaziabad' 'Hyderabad' 'Vasai-Virar' 'Kanpur' 'Patna'
 'Faridabad' 'Delhi' 'Unknown']
--------------------------------------------------
Unique values in Profession:
['Student' 'Civil Engineer' 'Architect' 'UX/UI Designer'
 'Digital Marketer' 'Content Writer' 'Educational Consultant' 'Teacher'
 'Manager' 'Chef' 'Doctor'

In [73]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27901 entries, 0 to 27900
Data columns (total 18 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   id                                     27901 non-null  int64  
 1   Gender                                 27901 non-null  object 
 2   Age                                    27901 non-null  float64
 3   City                                   27901 non-null  object 
 4   Profession                             27901 non-null  object 
 5   Academic Pressure                      27901 non-null  float64
 6   Work Pressure                          27901 non-null  float64
 7   CGPA                                   27901 non-null  float64
 8   Study Satisfaction                     27901 non-null  float64
 9   Job Satisfaction                       27901 non-null  float64
 10  Sleep Duration                         27901 non-null  object 
 11  Di

**Lưu thành file .csv sau khi tiền xử lý**

In [ ]:
X.to_csv("student_depression_dataset.csv", index=False)

## **Khám phá dữ liệu**

Bảng bên dưới mô tả chi tiết ý nghĩa của các thuộc tính trong bộ dữ liệu **`Student Depression Dataset`**


<center>

| **Tên thuộc tính**                        | **Mô tả** |
|-------------------------------------------|-------------------------------------------------------------|
| **id**                                    | Mã định danh của sinh viên. |
| **Gender**                                | Giới tính của sinh viên. Giá trị có thể là 'Male' (Nam) hoặc 'Female' (Nữ). |
| **Age**                                   | Tuổi của sinh viên. |
| **City**                                  | Thành phố nơi sinh viên sinh sống. |
| **Profession**                            | Nghề nghiệp của sinh viên. Giá trị có thể là 'Student' (Sinh viên). |
| **Academic Pressure**                     | Mức độ áp lực học tập của sinh viên (từ 1 đến 5, với 1 là ít và 5 là rất nhiều). |
| **Work Pressure**                         | Mức độ áp lực công việc của sinh viên (từ 1 đến 5, với 1 là ít và 5 là rất nhiều). |
| **CGPA**                                  | Điểm trung bình tích lũy (CGPA) của sinh viên. |
| **Study Satisfaction**                    | Mức độ hài lòng với việc học của sinh viên (từ 1 đến 5). |
| **Job Satisfaction**                      | Mức độ hài lòng với công việc của sinh viên (từ 1 đến 5). |
| **Sleep Duration**                        | Thời gian ngủ trung bình trong một ngày của sinh viên. Các giá trị có thể là 'Less than 5 hours', '5-6 hours', '7-8 hours', hoặc hơn. |
| **Dietary Habits**                        | Thói quen ăn uống của sinh viên. Các giá trị có thể là 'Healthy' (Lành mạnh) hoặc 'Moderate' (Hợp lý). |
| **Degree**                                | Loại bằng cấp của sinh viên, ví dụ: 'B.Pharm', 'BSc', 'M.Tech', v.v. |
| **Have you ever had suicidal thoughts?**  | Câu trả lời về việc sinh viên đã từng có suy nghĩ tự tử hay chưa. Giá trị có thể là 'Yes' hoặc 'No'. |
| **Work/Study Hours**                      | Số giờ làm việc hoặc học tập mỗi ngày của sinh viên. |
| **Financial Stress**                      | Mức độ căng thẳng tài chính của sinh viên (từ 1 đến 5, với 1 là ít và 5 là rất nhiều). |
| **Family History of Mental Illness**      | Liệu gia đình của sinh viên có tiền sử bệnh tâm lý hay không. Giá trị có thể là 'Yes' hoặc 'No'. |
| **Depression**                            | Mức độ trầm cảm của sinh viên. Giá trị có thể là '0' (không trầm cảm) hoặc '1' (trầm cảm). |


</center>



**Nhận xét về bộ dữ liệu Nấm (tập feature X) sau khi xử lý**
- Dữ liệu còn *27901* dòng và *18* thuộc tính (cột).
- Kiểu dữ liệu bao gồm, int, float và object.
- Không có giá trị khuyết thiếu trong bất kỳ thuộc tính nào vì tất cả các cột đều có 27901 giá trị không null.

## **Các câu hỏi**

#### **Câu 1**: ...


**Thuộc tính được sử dụng**
1. ..
2. ..

**Các bước thực hiện**
1. 
2. 
3.
4. 
5.

**Nhận xét**

